### ISSUE TITLE
FSHarp.Core 4.2.2 has same assembly version as 4.2.1

### ISSUE THREAD

**Issue by forki** [Microsoft/visualfsharp#3403]
I was happy to see a new 4.2.2 version on nuget.org, but unfortunately it has same assembly number as package version 4.2.1. They both have assembly version 4.4.1.0.

This will confuse people and tools and assembly binding redirects have no chance to tell which is which.

I fear we need yet another update.

**Comment by vasily-kirichenko** [comment]
and please add release notes like "update asap as 4.2.1 has async deadlocks and 4.2.2 has wrong assembly version"

**Comment by forki** [comment]
on the plus side: it looks like the dead lock is gone. So that's a clearly a win

**Comment by BentTranberg** [comment]
tip: In my solutions I create unit tests not only to test the source, but also to verify that the development environment and the output from it meet certain requirements. Testing that version numbers of libraries and tools satisfy certain requirements is one example. Testing that certain manual updates are completed would be another example. Perhaps a unit test can avoid this happening in the future.

**Comment by KevinRansom** [comment]
@vasily-kirichenko.  We don't rev the assembly version.  It only revs on major or minor version releases of the compiler.  So when F# 4.3 is released the assembly version will become 4.3.0.0.  We may change that behavior in the future, but for now it's not.

**Comment by forki** [comment]
You should at least update the patch level. This is a serious problem if we can't distinguish both dlls.

**Comment by KevinRansom** [comment]
Well … the loader distinguishes fine too and treats them as different assemblies, possibly loading them twice and causing type mismatches.  The FileVersion is set to the exact build version, so you can use that to distinguish between them if you need to..

**Comment by forki** [comment]
I cant distinguish between them from redirects. I cant set up redirects to
force the new one.

Am 29.07.2017 21:57 schrieb "Kevin Ransom (msft)" <notifications@github.com

**Comment by KevinRansom** [comment]
You can’t anyway unless they are installed in the GAC. And using the GAC is probably not that good a plan anyway.

**Comment by forki** [comment]
Oh I could make sure that the program only runs if it finds the correct
dll. This would be better than silent hangs.

Am 29.07.2017 22:07 schrieb "Kevin Ransom (msft)" <notifications@github.com

**Comment by KevinRansom** [comment]
Read the FileVer field.

**Comment by ReedCopsey** [comment]
Bumping the assembly version at the patch level seems like it has a lot of advantages, and few (no?) disadvantages. 

@KevinRansom: is there any technical reason you want to keep the assembly version the same for different versions of releases? That just makes life, from the POV of a developer and as somebody who tries to support the community, a total pain. It's really kind of violating normal rules for "releases" compared to _every other library_ out there.

**Comment by KevinRansom** [comment]
We rev FSharp.Core versions infrequently due to the vast number of scenarios that get broken when versions change.

You may consider this release a servicing release, as the package version number indicates.

Kevin

**Comment by 0x53A** [comment]
What about the scenarios that get broken when the version doesn't change?

**Comment by KevinRansom** [comment]
0x53a, which ones are they ... that don't include the GAC?

**Comment by KevinRansom** [comment]
I am wondering if you folks want me to update FSharp.Core.4.1.*.  I realize that one hasn't been updated to match the vs2017 Desktop compiler release.

4.1.17 doesn't have the regression since it was built using the RTM release,  I didn't release 4.1.18 when I put out the 4.2.1 release.

Kevin

**Comment by ReedCopsey** [comment]
@KevinRansom From my POV, no on updating 4.1.* - my understanding was that was supposed to stay untouched at this point.

> We rev FSharp.Core versions infrequently due to the vast number of scenarios that get broken when versions change.

I'd be more curious here - updating a version of a library _shouldn't_ cause "vast number of scenarios" to break. Specifically, why would updating the patch version of the assembly cause issues?  That's pretty much standard practice for library authors to do...

**Comment by forki** [comment]
Yes please don't push 4.1.x - that one is fine. 
We only hoped to see a 4.2.3 with exactly same code as 4.2.2 but with incremented assembly version number.

**Comment by matthid** [comment]
Not trying to be the devils advocate here, but I think I'm on @KevinRansom on this one.

- I have seen several places where FSharp.Core versions are hardcoded, maintaining them is a lot harder when we update everytime.
- Drop in replacements are easier when the assembly version doesn't change, this is by the way even considered best practice in .net in general: https://stackoverflow.com/questions/3768261/best-practices-guidance-for-maintaining-assembly-version-numbers

In general the runtime is quite picky about changes in the assembly version in particular. For bugfix releases you might not want that.

We definitely can think about this in general (maybe in an rfc process?), but IMHO it probably was the right call here. If we want to change this we need to think about all the options we break (for example simple in place upgrades without recompiling and hardcoded stuff in the ecosystem) and the new sources of errors we might introduce... And what we get in return.

**Comment by 0x53A** [comment]
1) You can't force an app to take the correct version - __a visible hard error is better than a silent hang__. Normally I hardcode the version of FSharp.Core in the app.config. If I (for some reason) have the wrong version in the build output, my program won't start (which is good!).

2) I have to admit I didn't test the scenario, so this may be a non-issue.
As I understand it, the loader will prefer the GAC to the current directory.
So if any other setup on the machine installed the bad version into the GAC, my app, that has the correct version locally, will still load the bad one. Is my understanding on that correct?

I've seen  a few apps that just GAC'd everything they depend on during the setup, regardless of whether it is smart or not.

**Comment by ReedCopsey** [comment]
@matthid Note that SO answer isn't "best practice" as much as "somebody's opinion that was popular at the time" - even the author of that answer stated he now ties version back to source control versions, etc (see comments).

@0x53A Yes, that is correct about GAC: see https://docs.microsoft.com/en-us/dotnet/framework/deployment/how-the-runtime-locates-assemblies - GAC is probed before local path, so any application that installs the bad 4.2.1 nuget version of this assembly into GAC will basically "break" that system for everybody unless the assembly version is bumped.

**Comment by KevinRansom** [comment]
The fix for that is to replace  the GACed version with the update.  And perhaps tell the authors to get out of the GAC.

Depend on package management for your build, the package dependencies say all of the right things.  I expect we may have a different policy on CoreClr versions when that becomes a thing, but desktop versions are likely to fit this mold for quite a while.

Perhaps you have forgotten the nightmare that was FSharp.Core versioning back in the FSharp 3.0 - 3.1 time frame.  The approach we currently take, is the one we have done for a long time, the versioning issues are known and explicable.

**Comment by 0x53A** [comment]
WE don't install anything in the GAC - my fear is another, completely unrelated app poisoning the system. I'm currently thinking of adding the fixed version of FSharp.Core to the GAC in my setup, just to work around this.

I think a lot of the issues in the olden times were just because best practices (app-local deployment of FSharp.Core, always have a BindingRedirect in the app.config) were not followed. Now, especially with paket and ``redirects:force``, but also with nuget and ``AutogenerateBindingRedirects``, this _should_ be a solved issue.

--------------

In the end it is your call.

**Comment by matthid** [comment]
@ReedCopsey

The comment is

>  This is quite an old post. Whilst most of it I still stand by there are some things that I would do different. NuGet versioning has changed how I do things and I do use Git tags for successful builds but at the end of the day the version number on the assembly should tie back to the build version on the build server and the tag version in source control. 

I pretty much agree on that. But you can use FileVersion for that (as the posts itself suggests). In no word he speaks about the assemblyversion attribute here.

Only tried to point out that there ARE real reason to NOT increase the version, as the discussion felt a bit one sided.

I have a lot experience with the runtime and assembly loading because of RazorEngine and my work on FAKE. It just is my feeling that it was the right decision, but I cannot prove it besides what I have already tried to point out. 

All I'm really saying is that this stuff is more complicated that it looks on the first sight. I agree that netcore might make this easier.

Can somebody point out the actual problems (besides msbuild being stupid and outdated gac) we have now?
You can in fact consider the gac thing as an advantage because you can patch existing apps (and people have used versioning like this before).

**Comment by KevinRansom** [comment]
The GAC problem identified by @0x53A is a very real one.  However, it is not a new problem it has existed for years,we have shipped updated FSharp.Core.dlls a bunch of times without revving the version number, we know how to deal with it.

We got out of the GAC in VS2015, that was one of my most favorite milestones.

It is worth considering how you want your library that you compiled against 4.2.2 FSharp.Core.dll without relying on the API that hangs to behave if it is loaded into VS-FSI on vs2017 RTM?   I believe it will fail to run if we used a revised version number and we absolutely want unpatched VS's to work fine with an updated FSharp.Core, it's why we try so hard to get compatibility right.

Versioning discussions are always difficult, and always unsatisfying.
